### PR TITLE
bindinfo: fix SPM doesn't work for CTE

### LINKED ISSUE #25505: SPM doesn't work for CTE
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

``` SQL
 CREATE TABLE `t` (
  `a` int(11) DEFAULT NULL,
  `b` int(11) DEFAULT NULL,
  KEY `b` (`b`)
);
set @@tidb_capture_plan_baselines='on';
with cte as (select * from t where b=4) select * from t;
with cte as (select * from t where b=6) select * from t;
with cte as (select * from t where b=7) select * from t;
show global bindings;
```

### 2. What did you expect to see? (Required)
show the binding.
### 3. What did you see instead (Required)
None.
### 4. What is your TiDB version? (Required)
master

### PR DESCRIPTION
**Pull request by zhaoxugang**
### What problem does this PR solve?

Issue Number: close #25505 

Problem Summary:

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:

How it Works:

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

- bindinfo: fix SPM doesn't work for CTE

### CODE REVIEW

**Comment by zhaoxugang** [comment]
@XuHuaiyu PTAL

**Comment by guo-shaoge** [comment]
> @XuHuaiyu PTAL

Guess it's better to ask help from planner-reviewers.

**Review by winoros — COMMENTED** [review]
Seems ok to me.
But it's better to add a test case that there're multiple WITH statement and multiple SELECT statement in its main body

**Comment by zhaoxugang** [comment]
\rebuild

**Comment by zhaoxugang** [comment]
> Seems ok to me.
> But it's better to add a test case that there're multiple WITH statement and multiple SELECT statement in its main body

done

File: bindinfo/handle.go [hunk 1]
```diff
@@ -704,7 +704,11 @@ func GenerateBindSQL(ctx context.Context, stmtNode ast.StmtNode, planHint string
 		bindSQL = bindSQL[updateIdx:]
 		return strings.Replace(bindSQL, "UPDATE", fmt.Sprintf("UPDATE /*+ %s*/", planHint), 1)
 	case *ast.SelectStmt:
+		withIdx := strings.Index(bindSQL, "WITH")
 		selectIdx := strings.Index(bindSQL, "SELECT")
+		if withIdx != -1 && withIdx < selectIdx {
+			selectIdx = withIdx
```
Comment (winoros) [inline bindinfo/handle.go:710]: Oh, raise up a question that should there be `withIdx+selectIdx`?

Comment (zhaoxugang) [reply bindinfo/handle.go:710]: like `SELECT * FROM t GROUP BY a WITH ROLLUP;`?

File: bindinfo/handle.go [hunk 2]
```diff
@@ -704,7 +704,11 @@ func GenerateBindSQL(ctx context.Context, stmtNode ast.StmtNode, planHint string
...
+		}
 		// Remove possible `explain` prefix.
 		bindSQL = bindSQL[selectIdx:]
 		return strings.Replace(bindSQL, "SELECT", fmt.Sprintf("SELECT /*+ %s*/", planHint), 1)
```
Comment (eurekaka) [inline bindinfo/handle.go:714]: The hint would be added to the select keyword in the cte definition, seems like that it cannot take effect for the main query block.

Comment (zhaoxugang) [reply bindinfo/handle.go:714]: it will be effect on sql like `with cte as (select * from t where b=7) select * from cte left join t1 on cte.a=t1.id;`

Comment (wjhuang2016) [reply bindinfo/handle.go:714]: @zhaoxugang For example, `with cte as (select * from t) select * from cte, t1 use index(a)` will be restored to `with cte as select /*+ use_index(t1, a) */ * from t  select * from cte, t1`. It's wrong

Comment (zhaoxugang) [reply bindinfo/handle.go:714]: @wjhuang2016 I try this case,but got  the result:
+-----------------------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------+------------+--------+-------------------------+-------------------------+---------+-----------+---------+
| Original_sql                                                                            | Bind_sql                                                                                                                                                   | Default_db | Status | Create_time             | Update_time             | Charset | Collation | Source  |
+-----------------------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------+------------+--------+-------------------------+-------------------------+---------+-----------+---------+
| with `cte` as ( select * from `test` . `t` ) select * from ( `cte` ) join `test` . `t1` | WITH `cte` AS (SELECT /*+ use_index(@`sel_1` `test`.`t1` `a`), hash_join(@`sel_1` `test`.`t1`)*/ * FROM `test`.`t`) SELECT * FROM (`cte`) JOIN `test`.`t1` | test       | using  | 2021-07-10 15:04:27.388 | 2021-07-10 15:04:27.388 |         |           | capture |
+-----------------------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------+------------+--------+-------------------------+-------------------------+---------+-----------+---------+

Comment (wjhuang2016) [reply bindinfo/handle.go:714]: @zhaoxugang No, the hint is wrong. It doesn't work for the main select.

Comment (zhaoxugang) [reply bindinfo/handle.go:726]: > @zhaoxugang No, the hint is wrong. It doesn't work for the main select.

done

File: bindinfo/handle.go [hunk 3]
```diff
@@ -704,14 +705,22 @@ func GenerateBindSQL(ctx context.Context, stmtNode ast.StmtNode, planHint string
...
-			selectIdx = withIdx
+		var selectIdx int
+		if n.With != nil {
+			var withSb strings.Builder
+			withIdx := strings.Index(bindSQL, "WITH")
+			restoreCtx := format.NewRestoreCtx(format.RestoreStringSingleQuotes|format.RestoreSpacesAroundBinaryOperation|format.RestoreStringWithoutCharset|format.RestoreNameBackQuotes, &withSb)
+			restoreCtx.DefaultDB = defaultDB
+			n.With.Restore(restoreCtx)
+			withEnd := withIdx + len(withSb.String())
+			tmp := strings.Replace(bindSQL[withEnd:], "SELECT", fmt.Sprintf("SELECT /*+ %s*/", planHint), 1)
```
Comment (wjhuang2016) [inline bindinfo/handle.go:720]: The hint may be in the `with` clause, for example, `with cte as (select * from t use index(idx)) select * from cte`.

Comment (wjhuang2016) [reply bindinfo/handle.go:720]: For the hint in the with clause, we can handle them like:
`with cte as (select b+1 from t use index(expression_index_5)) select * from cte;` 
to `with cte as (select b+1 from t) select /*+ use_index(@sel_2 t expression_index_5) */ * from cte;`

Please make sure the that query block is correct.

Comment (zhaoxugang) [reply bindinfo/handle.go:720]: > For the hint in the with clause, we can handle them like:
> `with cte as (select b+1 from t use index(expression_index_5)) select * from cte;`
> to `with cte as (select b+1 from t) select /*+ use_index(@sel_2 t expression_index_5) */ * from cte;`
> 
> Please make sure the that query block is correct.

it's ok, PTAL @wjhuang2016

File: server/tidb_test.go [hunk 4]
```diff
@@ -957,7 +957,7 @@ func (ts *tidbTestSerialSuite) TestTLS(c *C) {
 	c.Assert(stats, HasKey, "Ssl_server_not_after")
 	c.Assert(stats, HasKey, "Ssl_server_not_before")
 	c.Assert(stats["Ssl_server_not_after"], Equals, serverCert.NotAfter.Format("Jan _2 15:04:05 2006 MST"))
-	c.Assert(stats["Ssl_server_not_before"], Equals, serverCert.NotBefore.Format("Jan _2 15:04:05 2006 MST"))
+	c.Assert(stats["Ssl_server_not_before"], Equals, serverCert.NotBefore.Format("Jan 2 15:04:05 2006 MST"))
```
Comment (wjhuang2016) [inline server/tidb_test.go:960]: This line shouldn't be changed.

**Review by wjhuang2016 — COMMENTED** [review]
``` sql
CREATE TABLE `t` (
  `a` int(11) DEFAULT NULL,
  `b` int(11) DEFAULT NULL,
  KEY `idx_a` (`a`),
  KEY `idx_b` (`b`)
);
select 1 union (with cte as (select * from t use index(idx_a)) select a from cte);
```
Fail to capture this query.

**Comment by wjhuang2016** [comment]
> ```sql
> CREATE TABLE `t` (
>   `a` int(11) DEFAULT NULL,
>   `b` int(11) DEFAULT NULL,
>   KEY `idx_a` (`a`),
>   KEY `idx_b` (`b`)
> );
> select 1 union (with cte as (select * from t use index(idx_a)) select a from cte);
> ```
> 
> Fail to capture this query.

It seems that capture doesn't support UNION yet. Let other PR handle it.

**Comment by zhaoxugang** [comment]
@Reminiscent PTAL

File: executor/adapter.go [hunk 5]
```diff
@@ -1095,9 +1095,6 @@ func getEncodedPlan(sctx sessionctx.Context, p plannercore.Plan, genHint bool, n
 	}
 	if genHint {
 		hints := plannercore.GenHintsFromPhysicalPlan(p)
-		if n != nil {
-			hints = append(hints, hint.ExtractTableHintsFromStmtNode(n, nil)...)
-		}
```
Comment (Reminiscent) [inline executor/adapter.go:1100]: Why should we remove these lines?

**Review by Reminiscent — APPROVED** [review]

File: executor/adapter.go (hunk 5 above)
Comment (zhaoxugang) [reply executor/adapter.go:1100]: I think it is enough to gather hints from the physical plan .

Comment (Reminiscent) [reply executor/adapter.go:1100]: Please add some tests about table hints in the CTE statement. @zhaoxugang

**Review by winoros — APPROVED** [review]

**Review by wjhuang2016 — APPROVED** [review]

**Pull request merged**